File: src/Lib/Conf/Conf.ts
/*
 * Copyright (c) 2023 Kiyozz~WK~WushuLate.
 *
 * All rights reserved.
 *
 */
import { join } from '@tauri-apps/api/path'
import { catchErr } from 'App/Lib/TsResults'
import deepmerge from 'deepmerge'
import { Ok, type Result } from 'ts-results'
import { type PartialDeep } from 'type-fest'
import { type ConfOptions } from 'App/Lib/Conf/ConfOptions'
import { type Conf } from 'App/Lib/Conf/ConfZod'
import { isConfFileExists, readConfigFileJson } from 'App/Lib/Conf/ReadConfigFile'
import { writeConfigFile } from 'App/Lib/Conf/WriteConfigFile'

export const confFileName = 'conf.json'

/**
 * Write the default conf if the conf file doesn't exist
 *
 * @param options
 */
const writeDefaultConfIfFileNotExists = async (options: ConfOptions<Conf>): Promise<Result<void, Error>> => {
  const notNeedWrite = (await isConfFileExists()).unwrap()

  if (notNeedWrite) {
    return Ok(undefined)
  }

  return await writeConfigFile(options.defaults)
}

const readConfigOrUseDefaultConfig = async (): Promise<Result<Conf, Error>> => {
  return await catchErr(async () => await readConfigFileJson())
}

const defaultOptions: ConfOptions<Conf> = {
  projectVersion: '2023.1',
  confName: confFileName,
  defaults: {
    game: {
      type: 'Skyrim SE',
      path: '',
    },
    compilation: {
      compilerPath: await join('Papyrus Compiler', 'PapyrusCompiler.exe'),
      flag: 'TESV_Papyrus_Flags.flg',
      concurrentScripts: 15,
      output: await join('Data', 'Scripts'),
    },
    tutorial: {
      telemetry: true,
      settings: true,
    },
    mo2: {
      use: false,
      instance: undefined,
      output: await join('overwrite', 'Scripts'),
      modsFolderRelativeToInstance: 'mods',
    },
    telemetry: {
      use: true,
    },
    theme: 'system',
    locale: 'fr',
    misc: {
      documentation: {
        reminder: true,
      },
      drawerOpen: false,
    },
    logLevel: 'error',
  },
}

export const readConfig = async () => await readConfigOrUseDefaultConfig()

export const writeConfig = async (partialConfig: PartialDeep<Conf>): Promise<Result<Conf, Error>> => {
  const conf = await readConfig()
  const finalConf = conf.map((conf) => deepmerge(conf as Partial<Conf>, partialConfig))

  if (finalConf.err) return finalConf

  await writeConfigFile(finalConf.val)

  return finalConf
}

await writeDefaultConfIfFileNotExists(defaultOptions)

File: src-tauri/src/pca/conf.rs
/*
 * Copyright (c) 2023 Kiyozz~WK~WushuLate.
 *
 * All rights reserved.
 *
 */

use crate::pca::{data::ConfPath, logs::LogsState};
use log::{trace, LevelFilter};
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::Path;
use tauri::api::file;
use tauri::{Manager, PathResolver, Window, Wry};

/// This is the application conf.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Conf {
    pub game: Game,
    pub compilation: Compilation,
    pub tutorial: Tutorial,
    pub mo2: Mo2,
    pub telemetry: Telemetry,
    pub theme: String,
    pub locale: String,
    pub misc: Misc,
    pub log_level: String,
}

impl Default for Conf {
    /// Create a default Conf. Used by the application menu to reset the application conf.
    fn default() -> Self {
        let compiler_path = Path::new("Papyrus Compiler");
        let compilation_output = Path::new("Data");
        let mo2_output = Path::new("overwrite");

        Conf {
            game: Game {
                type_field: "Skyrim SE/AE".to_string(),
                path: "".to_string(),
            },
            compilation: Compilation {
                compiler_path: compiler_path
                    .join("PapyrusCompiler.exe")
                    .to_str()
                    .unwrap()
                    .to_string(),
                flag: "TESV_Papyrus_Flags.flg".to_string(),
                concurrent_scripts: 15,
                output: compilation_output
                    .join("Scripts")
                    .to_str()
                    .unwrap()
                    .to_string(),
            },
            tutorial: Tutorial {
                settings: false,
                telemetry: true,
            },
            mo2: Mo2 {
                use_field: false,
                instance: None,
                output: mo2_output.join("Scripts").to_str().unwrap().to_string(),
                mods_folder_relative_to_instance: "mods".to_string(),
            },
            telemetry: Telemetry { use_field: false },
            theme: "system".to_string(),
            locale: "fr".to_string(),
            misc: Misc {
                documentation: Documentation { reminder: true },
                drawer_open: false,
            },
            log_level: "error".to_string(),
        }
    }
}

impl Conf {
    pub fn reset(resolver: &PathResolver) -> Result<Conf, crate::pca::Error> {
        let conf = Conf::default();

        conf.save(resolver).map_err(crate::pca::Error::from)?;

        Ok(conf)
    }

    pub fn log_level_filter(&self) -> LevelFilter {
        match self.log_level.as_str() {
            "trace" => LevelFilter::Trace,
            "debug" => LevelFilter::Debug,
            "info" => LevelFilter::Info,
            "warn" => LevelFilter::Warn,
            "error" => LevelFilter::Error,
            _ => LevelFilter::Off,
        }
    }

    pub fn get(resolver: &PathResolver) -> Result<Self, crate::pca::Error> {
        let settings_file =
            file::read_string(resolver.app_conf()).map_err(crate::pca::Error::from)?;
        let json: Result<Conf, _> = serde_json::from_str(settings_file.as_str());

        json.map_err(crate::pca::Error::from)
    }

    pub fn save(&self, resolver: &PathResolver) -> Result<(), crate::pca::Error> {
        let json = serde_json::to_string_pretty(self).expect("Failed to serialize settings");

        fs::write(resolver.app_conf(), json).map_err(crate::pca::Error::from)
    }

    pub fn emit_reset(&self, window: &Window<Wry>) {
        window
            .emit("pca://conf_reset", self)
            .expect("Failed to emit reset_conf event");
    }
}

#[tauri::command]
pub fn on_conf_reset(window: Window<Wry>, logs_state: tauri::State<&LogsState>) {
    trace!("reset conf file");

    match &logs_state.logs {
        Ok(logs) => {
            let path_resolver = &window.app_handle().path_resolver();
            let new_conf = Conf::reset(path_resolver).expect("Failed to reset settings");

            logs.on_conf_reset(path_resolver)
                .expect("Cannot logs.on_conf_reset after Conf::reset from on_conf_reset command");

            trace!("conf reset");

            new_conf.emit_reset(&window);
        }
        Err(err) => {
            trace!("failed to reset conf file: {:#?}", err);
        }
    }
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Game {
    #[serde(rename = "type")]
    pub type_field: String,
    pub path: String,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Compilation {
    pub compiler_path: String,
    pub flag: String,
    pub concurrent_scripts: i64,
    pub output: String,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Tutorial {
    pub settings: bool,
    pub telemetry: bool,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Mo2 {
    #[serde(rename = "use")]
    pub use_field: bool,
    pub output: String,
    pub mods_folder_relative_to_instance: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    pub instance: Option<String>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Telemetry {
    #[serde(rename = "use")]
    pub use_field: bool,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Misc {
    pub drawer_open: bool,
    pub documentation: Documentation,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Documentation {
    pub reminder: bool,
}
